tsdb/wlog/watcher_test.go

```diff
@@ -755,36 +752,43 @@ func TestRun_AvoidNotifyWhenBehind(t *testing.T) {
 
 			w, err := NewSize(nil, nil, wdir, segmentSize, compress)
 			require.NoError(t, err)
-			var wg sync.WaitGroup
 			// Generate one segment initially to ensure that watcher.Run() finds at least one segment on disk.
 			require.NoError(t, generateWALRecords(w, 0, seriesCount, samplesCount))
 			w.NextSegment() // Force creation of the next segment
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				for i := 1; i < segmentsToWrite; i++ {
-					require.NoError(t, generateWALRecords(w, i, seriesCount, samplesCount))
-					w.NextSegment()
-				}
-			}()
 
+			// Set up the watcher and run it in the background.
 			wt := newWriteToMock(time.Millisecond)
 			watcher := NewWatcher(wMetrics, nil, nil, "", wt, dir, false, false, false)
+			watcher.setMetrics()
 			watcher.MaxSegment = segmentsToRead
 
-			watcher.setMetrics()
-			startTime := time.Now()
-			err = watcher.Run()
-			wg.Wait()
-			require.Less(t, time.Since(startTime), readTimeout)
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				startTime := time.Now()
+				err = watcher.Run()
+				require.NoError(t, err)
```

Comment:
It's not allowed to call t.FailNow on a background goroutine.
Perhaps use golang.org/x/sync/errgroup and return the error?

Reply:
`require.NoError` calls `t.FailNow` under the hood, so I think Bryan's comment here is still relevant

Reply:
Yes, I'm aware of the "issue," but I've always ignored it and have never encountered a false "test passed" because of it. As long as the test fails when necessary, I'm not too concerned. That said, I'm okay with implementing a cleaner solution.

We already use this in other places (Go really just needs to support it; it's simpler and more natural).

That's why `testifylint`'`go-require` that help catch some of them is disabled: https://github.com/prometheus/prometheus/blob/349068ad3ed81d32aa78b8a66da9a6fbec1ac333/.golangci.yml#L152-L155C9

Reply:
done, and adjusted some comments.